pm: domains: Initial power domain support

ceolin opened the pull request:
Add support for power domains on Zephyr. Power domains are implemented as simple devices so they can use the existent Zephyr API, for resume and suspend sync and async and also reference count.  

- Power domain drivers are responsible for notifying child devices of state changes themselves.
- Custom drivers can be notified and handle transition failures if they wish.
- Power domain transitions are integrated into the existing PM actions, instead of a new set of actions that can occur 
 asynchronously.

The pm subsystem will ensure that domains are resumed before and suspended after devices using them. For device runtime power management, every time the device is got or released the same actions is done to the domain it belongs. 

As domains are implemented as simple devices, it is totally acceptable a domain belongs to another domain.

New changes:

 - Added a  new device action: `PM_DEVICE_ACTION_TURN_ON`
   - `PM_DEVICE_TURN_ON` and `PM_DEVICE_TURN_OFF` actions are used to notifiy devices when the power domain they belong is resuming or suspending.
 - The power domain attribute was moved to `pm.yaml`
 - device runtime is calling `pm_device_state_set` instead of `pm->action_cb` becasuse of the power domain logic that is handled in that function

An overall idea of two devices under a power domain doing runtime power management.
![pd](https://user-images.githubusercontent.com/882919/142272710-68cf1101-2c0c-47e8-8f7d-c956e97ee6b8.png)

Commit:
pm: Add power domain infra structure

Add support for power domains on Zephyr. Power domains are implemented
as simple devices so they can use the existent Zephyr API, for resume
and suspend sync and async and also reference count.

The pm subsystem will ensure that domains are resumed before and
suspended after devices using them. For device runtime power
management, every time the device is got or released the same actions
is done to the domain it belongs.

As domains are implemented as simple devices, it is totally acceptable
a domain belongs to another domain.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
docs: pm: Add power domain information

Document power domains on Zephyr.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>
Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

teburd reviewed (commented):
The power domain support looks pretty good to me, one thing that isn't clear here is how a domain that is depended on by many devices deals with the request to suspend when not all devices are yet suspended, I may have overlooked where that happens though.

JordanYates reviewed (changes requested):
I have several concerns/questions about the overall approach.

Why should we introduce a new scheme for power domains when we already have the `regulator-fixed`/`regulator-sync` devices, and the corresponding DT properties in `power.yaml`?

This PR doesn't add support for even the simplest of setups that would need to use this API, i.e. a GPIO regulator with a device on it.

There is no support for notifying other devices on a power domain that they have suddenly powered on. There are plenty of devices that boot up in silly modes that draw milliamps, which is not compatible with low power operation.

Because there is no functionality beyond calling `pm_device_get` on any `power-domain` phandle attached to a device, any iteration code to do the above would need to be duplicated in all devices implementing `power-domain`.

Personally I think that power domain support needs to be a new API (`pm_device_power_get` or similar), that has explicit behavior in terms of turning on nested domains and notifying supported devices that they have had power applied/removed.

JordanYates commented:
> Because they a power domain is not necessarily a regulator.
```
 - Power Domain - Electronic circuit that is supplied its input power by the
                   output power of a regulator, switch or by another power
                   domain.
```
A switch is equivalent in all ways to a regulator that doesn't change the voltage between input and output.
I see that the code you implemented also uses this for internal SOC dependencies, but I'm not convinced that SOC peripheral drivers need to be using this API (to me that feels much more like the clock_control API).

Are there other examples of things that aren't regulators or switches that would be using this API?

> Can you elaborate more ?If you declare a domain that controls the GPIO regulator and the device uses this domain the workflow is correct isn't ?

Maybe, but my point is that a PR implementing "initial power domain support" doesn't add support for the simplest use case.

> I particularly don't see a good reason to have a new API, couldn't just the subsystem ensure that when a power domain change status your dependents are notified ? And by notified it can be as simple as call dev->pm_control with the right action.

The advantage of a new API is that if the API itself is walking the dependency/support tree and notifying devices as it goes along, you are no longer depending on individual drivers to forward the requests around.

Without any example implementations of your `power-domain`, it is hard to explore the requirements and whether a separate API makes sense. The TI internal power domains are not general enough for this purpose (it is guaranteed that there are no chained domains, and other devices are not being notified of state changes)

gmarull commented:
I think that to make progress, it would be good to compare in detail this proposal with Linux PM domains.

ceolin commented:
> > > I think that to make progress, it would be good to compare in detail this proposal with Linux PM domains.
> > 
> > 
> > This is not Linux, use cases are different, e.g. Linux allows attach / detach devices at runtime while on Zephyr everything is static. Though we can make a comparison at some points, also I've looked TI HAL that implements power domains, can include on it.
> > I just don't think that compare with Linux and implement what is done there is what we need, rather would be better having real use cases where this would be helpful.
> 
> I'm not saying we need to follow Linux, but since it seems they have PM domains implemented, I think it's worth comparing to what they have. We may just need a subset of what they offer, but it's worth understanding _which_ subset and _why_.

that is fair, just the `_why_` that shouldn't come from there but from our needs.

teburd commented:
> > The power domain support looks pretty good to me, one thing that isn't clear here is how a domain that is depended on by many devices deals with the request to suspend when not all devices are yet suspended, I may have overlooked where that happens though.
> 
> which situation ? When the kernel is suspending devices, the domain will be suspended only after all devices depending on it be suspended. This is currently only based on the device initialization level, but it needs to be changed to used the "supported" feature that was recently removed.
> 
> When device runtime is used, the reference count of the domain is incremented every time a device that needs it is in use and decreased when the device is not longer needed.

That makes perfect sense to me, thanks

gmarull commented on subsys/pm/device.c:
```diff
@@ -24,12 +24,26 @@ const char *pm_device_state_str(enum pm_device_state state)
 	int ret;
 	enum pm_device_action action;
 	struct pm_device *pm = dev->pm;
+	bool domain = atomic_test_bit(&dev->pm->flags,
```
are atomics required? isn't a const attribute to be a PD?

ceolin replied:
nope, this is just a flag. The field `domain` is used to inform a a device which domain its belong.
Just to be clear, the flag is used to inform that device itself is a power domain.

ceolin replied:
I don't know if it was clear, but using atomic is necessary because the flags type is `atomic_t`. I'm leveraging this variable to set a domain instead of an additional boolean into the struct.

gmarull commented on subsys/pm/device.c:
```diff
@@ -44,6 +58,16 @@ int pm_device_state_set(const struct device *dev,
+		 * about it
+		 */
+		if (domain) {
+			(void)device_supported_foreach(dev,
+				      domain_supported_cb,
+				      (void *)PM_DEVICE_ACTION_POWER_DOMAIN_OFF);
```
I wonder if devices should enable notification (via a flag), since many probably don't care. cc: @JordanYates ?

ceolin replied:
to avoid the function call ? if they don't want they can ignore (don't handle) the action. Just wondering if it is not better to keep it simpler.

JordanYates replied:
I agree with @ceolin, explicitly ignoring the action in the callback is simpler than adding a reason for the callback to not be run.

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -63,7 +63,7 @@ static int runtime_suspend(const struct device *dev, bool async)
 		(void)k_work_schedule(&pm->work, K_NO_WAIT);
 	} else {
 		/* suspend now */
-		ret = pm->action_cb(pm->dev, PM_DEVICE_ACTION_SUSPEND);
```
I think we should not use `pm_device_state_set` for multiple reasons:

- It uses states, not actions (`pm_device_state_set` is still available for legacy reasons really, so let's not use in new code unless there is a good reason or no other way)
- `pm_device_state_set` does checks already done by the runtime PM

Note that I see another issue: now devices using runtime PM are ignored by system PM, but, domains will not be ignored by system. If system then suspends a domain, what will happen to a devices attached to the domain using runtime PM?

ceolin replied:
The first one is something that makes sense and has to be addressed. Will add a commit doing it and ensure that the action function is used.
The second reason,  I need to think, I don't think we should have the same logic scattered across the code.

ceolin replied:
Regarding device runtime power management, I think it is hard to automate anything here, I believe if the application decided to enable runtime pm for a device it is its responsibility do it for the the domain if it is worried about system pm or to  set a constraint in the system to not sleep and not suspend devices. 

Automatically enable runtime power management on domain can be problematic because the domain may already being used by another device.

What we can do is automatically get / put the domain according with the device usage, since per conception if a device is under the domain it needs the domain active to be active and that is what it does.

gmarull commented on subsys/pm/device.c:
```diff
@@ -24,12 +24,26 @@ const char *pm_device_state_str(enum pm_device_state state)
+	enum pm_device_action action = (enum pm_device_action)context;
+
+	if (pm->action_cb != NULL) {
+		(void)pm->action_cb(dev, action);
```
are all errors ignored on purpose? maybe just `-ENOTSUP` should be ignored?

gmarull replied:
I think ignoring errors here is a dangerous practice: system will be left in a state where certain devices will not be operational without any notice.

gmarull commented on doc/guides/pm/device.rst:
```diff
@@ -134,3 +134,68 @@ later calling :c:func:`pm_device_wakeup_enable`.
+
+.. code-block:: devicetree
+
+		gpioDomain: gpioDomain@4 {
```
`gpio_domain` (no camelCase) by convention

gmarull commented on tests/subsys/pm/device_wakeup_api/src/main.c:
```diff
@@ -9,12 +9,55 @@
+static const struct device *gpio_dev, *domain_dev;
 static uint8_t sleep_count;
 
+static int domain_gpio_init(const struct device *dev)
```
I'm not sure if mixing PM domain tests with wakeup API tests is a good idea.

ceolin replied:
removed !

gmarull commented on include/pm/device.h:
```diff
@@ -142,6 +144,9 @@ struct pm_device {
 				(DT_PROP_OR(node_id, wakeup_source, 0)),\
 				(0)) << PM_DEVICE_FLAG_WS_CAPABLE),	\
+		.domain = COND_CODE_1(DT_NODE_HAS_PROP(node_id, power_domain),	\
+			(DEVICE_GET(Z_DEVICE_DT_DEV_NAME(DT_PROP(node_id, \
```
can `DEVICE_DT_GET_OR_NULL(DT_PHANDLE(node_id, power_domain))` can be used here?

ceolin replied:
Yep, this macro was introduce after I've written it. The macro make it simpler indeed.

gmarull replied:
not resolved

ceolin replied:
it was in the commit making it optional.

gmarull commented on include/pm/device.h:
```diff
@@ -113,6 +113,8 @@ struct pm_device {
 	enum pm_device_state state;
 	/** Device PM action callback */
 	pm_device_action_cb_t action_cb;
+	/** Power Domain it belongs */
+	const struct device *domain;
```
nit: domains could be optional (to allow for resource optimization if not required)

ceolin replied:
yep, that can be done

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -94,9 +94,20 @@ static void runtime_suspend_work(struct k_work *work)
 	__ASSERT(ret == 0, "Could not suspend device (%d)", ret);
 }
 
+#define PM_CHECK_RETURN(_ret) \
+	((ret == 0) || (ret == -ENOTSUP) || (ret == -ENOSYS))
```
are all these error codes meaningfull for domains? `-ENOSYS` surely not, `-ENOTSUP` probably not if we require domains to implement suspend/resume.

ceolin replied:
That is the thing, power domains is just a regular device and the action callback is not enforced.  Consequently, if a power domain does not implement an action callback, it will never be suspended and will return `-ENOSYS`. If that that makes sense (having a power domain that does not implement action cb) is other question.

gmarull replied:
I think there should be a contract here: a power domain needs to implement suspend and resume. Adding this check will mask implementation issues.

ceolin replied:
Fair enough. Lets enforce domains to implement the action callback and handle SUSPEND/RESUME

JordanYates commented:
@ceolin With #40418 being merged, the revert commits in this PR can be removed

Commit:
pm: device: Make power domain optional

Add a Kconfig symbol to enable/disable power domain on Zephyr.
Disabling power domain save some memory / space.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: pm: Add a power domain test

Add simple test to exercise power domain behavior
with device runtime.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

teburd reviewed (approved):
This seems good to me, the wakeup test mixed with the power domain test I understand the reasoning for while also understanding the comment from @gmarull. Personally its not a deal breaker to me as there are intertwined features that need to be tested together, the best place to do so isn't really clear.

What would be great in maybe another patch (doesn't have to be from @ceolin) is excercising this API more in a real board setting with internal/external gates and devices involved. Maybe one of the nordic thingy's as suggest, would be a good candidate for a future showcase.

ceolin commented:
> This seems good to me, the wakeup test mixed with the power domain test I understand the reasoning for while also understanding the comment from @gmarull. Personally its not a deal breaker to me as there are intertwined features that need to be tested together, the best place to do so isn't really clear.
> 
> What would be great in maybe another patch (doesn't have to be from @ceolin) is excercising this API more in a real board setting with internal/external gates and devices involved. Maybe one of the nordic thingy's as suggest, would be a good candidate for a future showcase.

Agree, I don't have access to this board. The one I've used is one TI but the way power domain is implemented in that board requires bigger changes and is out of the scope of this pr IMHO.
The way it is proposed / implemented in this PR, power domains does not break current API and is tottaly optional not increasing memory footprint when disabled.

JordanYates commented on doc/glossary.rst:
```diff
@@ -75,6 +75,10 @@ Glossary of Terms
 
+   power domain
+      A power domain is a collection of gates powered by the same power
+      and ground supply. Power domains on are represented by :c:struct:`device`.
```
I don't think this description is strictly accurate. I could easily see multiple grounds being involved. The term "gates" also seems rather specific to SoC internals. Perhaps something along the lines of
```
is a collection of devices for which power is applied and removed collectively in a single action. 
```

ceolin replied:
I liked it.

JordanYates commented on dts/bindings/base/pm.yaml:
```diff
@@ -16,3 +16,13 @@ properties:
+      type: phandle
+      description: |
+
+        Power domain the device belongs.
```
belongs to.

JordanYates commented on dts/bindings/base/pm.yaml:
```diff
@@ -16,3 +16,13 @@ properties:
+
+        Power domain the device belongs.
+
+        The device will be notified when the power domain it belongs is either
```
belongs to

JordanYates reviewed (commented):
I like the general concepts presented here, and have used most of it as inspiration when updating my forks power domain support for v2.7.0.

I strongly believe this needs at least one example of controlling power domains external to the SoC in order to validate that the overall flow makes sense (Ideally with end device support). I know you have commented
> The way it is proposed / implemented in this PR, power domains does not break current API and is tottaly optional not increasing memory footprint when disabled.

but as soon as this is committed it becomes an API that people may start depending on, and decisions become harder to change. Additionally it would provide a reference implementation for people looking to update other devices to support power domains.

ceolin commented on include/pm/device.h:
```diff
@@ -73,6 +76,10 @@ enum pm_device_action {
 	PM_DEVICE_ACTION_TURN_OFF,
```
I completely ignored this action, it has to be accounted and a proper notification should be sent.
Thanks for pointing it.

ceolin commented on include/pm/device.h:
```diff
@@ -73,6 +76,10 @@ enum pm_device_action {
 	/** Force suspend. */
 	PM_DEVICE_ACTION_FORCE_SUSPEND,
+	/** Power domain on */
+	PM_DEVICE_ACTION_DOMAIN_ON,
```
You, mean `PM_DEVICE_ACTION_DOMAIN_TURN_ON|OFF` ?

JordanYates replied:
Not really, `PM_DEVICE_ACTION_RESUME` for example is using the verb `RESUME` to tell the device to move from suspended to active. `PM_DEVICE_ACTION_DOMAIN_ON` has no verb, it is not actually describing an action. `DOMAIN_ON` has already happened, it is not something you are asking the device to do.

gmarull replied:
I tend to agree with @JordanYates here. Actions tell the device what to do, so `PM_DEVICE_ACTION_DOMAIN_ON` doesn't specify any action. Maybe `PM_DEVICE_ACTION_HANDLE_PD_ON/OFF` or `PM_DEVICE_ACTION_HANDLE_DOMAIN_ON/OFF`?

ceolin replied:
Oo I see. I agree, we have this discussion some time ago and I just kept these names because nothing better had suggested.

gmarull commented on dts/arm/ti/cc13x2_cc26x2.dtsi:
```diff
@@ -45,6 +45,20 @@
 	};
 
 	soc {
+		serialDomain: serialDomain {
```
nit: by convention `serial_domain`

gmarull commented on dts/bindings/power-domain/ti,cc13xx-cc26xx-power-domain.yaml:
```diff
@@ -0,0 +1,15 @@
+include: power-domain.yaml
+
+properties:
+    resource-id:
+      type: int
+      required: true
+      description:
+        Power domain resource ID as defined in prcm.h (TI HAL).
```
would be nice to provide a dt-bindings header with pre-defined resource IDs (DT can't include HAL headers).

ceolin replied:
something like this: https://github.com/ceolin/zephyr/blob/pm/pd/dts/bindings/spi/xlnx,xps-spi-2.00.a.yaml#L20-L29 ?

Not sure if this helps here, since an integer value will also not being that meaningful. Also, it is pretty common set value directly in DT, all registers are arbitrarily set.

gmarull replied:
I mean to provide a header in `include/dts-bindings/` with pre-defined resource-ids.

So:
```
			/* PRCM_DOMAIN_PERIPH value */
			status = "okay";
			resource-id = <0x00000004>;
```

can be written as:
```
			status = "okay";
			resource-id = <PRCM_DOMAIN_PERIPH>;
```

gmarull commented on soc/arm/ti_simplelink/cc13x2_cc26x2/power.c:
```diff
@@ -238,3 +243,65 @@ bool pm_constraint_get(enum pm_state state)
+									\
+	PM_DEVICE_DT_INST_DEFINE(n, domain_cc13xx_cc26xx_pm_action);	\
+									\
+	DEVICE_DT_INST_DEFINE(n,					\
+		      domain_cc13xx_cc26xx_init,			\
+		      PM_DEVICE_DT_INST_REF(n),				\
+		      NULL, &domain_cc13xx_cc26xx_config_##n,		\
+		      PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,	\
+		      NULL);
```
non-blocking, but I think formatting should be like this here:

```c
	DEVICE_DT_INST_DEFINE(n, domain_cc13xx_cc26xx_init,		       \
			      PM_DEVICE_DT_INST_REF(n), NULL,		       \
			      &domain_cc13xx_cc26xx_config_##n,		       \
			      PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,\
			      NULL);
```

gmarull commented on tests/subsys/pm/device_wakeup_api/src/main.c:
```diff
@@ -9,12 +9,56 @@
+}
+
+PM_DEVICE_DT_INST_DEFINE(0, domain_gpio_pm_action);
+DEVICE_DT_DEFINE(DT_INST(0, power_domain), domain_gpio_init,
+		PM_DEVICE_DT_INST_REF(0), NULL, NULL,
+		PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+		NULL);
```
nit: align after `(`

gmarull commented on include/pm/device.h:
```diff
@@ -144,10 +151,13 @@ struct pm_device {
+							power_domain)), \
+				(0)) << PM_DEVICE_FLAG_POWER_DOMAIN),	\
+		.domain = DEVICE_DT_GET_OR_NULL(DT_PHANDLE(node_id,	\
+					power_domain)),			\
```
please fixup first commit adding `.domain = ...`

ceolin replied:
ok

gmarull commented on include/pm/device.h:
```diff
@@ -144,10 +151,13 @@ struct pm_device {
+				COND_CODE_1(DT_NODE_EXISTS(node_id),	\
+					    (DT_NODE_HAS_COMPAT(node_id,\
+							power_domain)), \
+				(0)) << PM_DEVICE_FLAG_POWER_DOMAIN),	\
```
note: it may be time to add a helper here, this is getting cluttered

gmarull commented on subsys/pm/device.c:
```diff
@@ -44,6 +59,16 @@ int pm_device_state_set(const struct device *dev,
+		 * about it
+		 */
+		if (domain) {
+			(void)device_supported_foreach(dev,
+			      domain_supported_cb,
+			      (void *)PM_DEVICE_ACTION_DOMAIN_OFF);
```
nit: why is error code ignored here?

ceolin replied:
The problem is what do with another error.

If the domain is resuming the action happens before notify devices, in this case, if a device returns an error we would need to suspend the domain again ? And notify all devices that were already notified ?

If the domain is supending, what to do, cancel the operation and notify devices that received the notification that the domain will continue active ?

Also, this will require all devices properly return -ENOTSUP or 0 when not handling these notifications. I don't know if initially we should be doing it, but I can change it though.

gmarull commented on subsys/pm/device.c:
```diff
@@ -70,6 +95,13 @@ int pm_device_state_set(const struct device *dev,
 
+	/* Power domain is ON, now notify all dependents about it */
+	if (domain && action == PM_DEVICE_ACTION_RESUME) {
+		(void)device_supported_foreach(dev,
+				      domain_supported_cb,
+				      (void *)PM_DEVICE_ACTION_DOMAIN_ON);
```
ditto

ceolin replied:
see reply above

gmarull reviewed (commented):
I basically have these concerns here:

- I think the modifications to `pm_device_action_run` will in the end perpetuate its usage for manual or custom PM solutions.

  - `pm_device_action_run` is used by system PM, however, nothing special is required there (domain is a regular device that will come before its children when resuming and after when suspending). TLDR; `pm_device_action_run` can be left as is and system PM will just work.
  - `pm_device_action_run` is introduced in the runtime PM API, however, it does extra work not needed by the runtime API. Why not add the domain changes there?

- Documentation should be improved to describe usages of power domains, provide an implementation example, usage guidelines, etc.
- There is now a single implementation (TI) and a really basic test, so I'm not 100% sure this solution will fit other use cases. It would be nice to try to solve another problem with this API (IMO it's a good practice when adding new APIs like this one). This of course can also be covered with more testing/docs when not possible.

ceolin commented:
;;; Type C-c C-c to include your comment
Oo I see. I agree, we have this discussion some time ago and I just kept these names because nothing better had suggested.

ceolin commented:
;;; Type C-c C-c to include your comment
The problem is what do with another error.

If the domain is resuming the action happens before notify devices, in this case, if a device returns an error we would need to suspend the domain again ? And notify all devices that were already notified ?

If the domain is supending, what to do, cancel the operation and notify devices that received the notification that the domain will continue active ?

Also, this will require all devices properly return `-ENOTSUP` or 0 when not handling these notifications. I don't know if initially we should be doing it, but I can change it though.

ceolin commented:
Oo I see. I agree, we have this discussion some time ago and I just kept these names because nothing better had suggested.

ceolin commented:
Oo I see. I agree, we have this discussion some time ago and I just kept these names because nothing better had suggested.

peter-mitsis reviewed (approved):
Being less familiar with power management, I might be missing something, but it looks good as far as I can tell.

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -92,6 +92,14 @@ static void runtime_suspend_work(struct k_work *work)
+	 */
+	if (pm->domain != NULL) {
+		(void)pm_device_runtime_put_async(pm->domain);
+	}
```
note: this is already called within a work queue context, so why async?

ceolin replied:
This gives the scheduler an opportunity to schedule another thread since this operation does not need to happen synchronously. Though, this is not something  critical, maybe just taking advantage that it is already running in the context of the workqueue and make it synchronously will keep things simpler.

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -147,6 +166,13 @@ int pm_device_runtime_put(const struct device *dev)
+	 */
+	if ((ret == 0) && dev->pm->domain != NULL) {
+		ret = pm_device_runtime_put(dev->pm->domain);
+	}
```
`dev->pm` -> `pm->`

ceolin replied:
I change it in 1fe5efe

gmarull commented on include/pm/device.h:
```diff
@@ -77,6 +80,12 @@ enum pm_device_action {
 	PM_DEVICE_ACTION_TURN_OFF,
 	/** Force suspend. */
 	PM_DEVICE_ACTION_FORCE_SUSPEND,
+	/** Power domain resume */
+	PM_DEVICE_ACTION_HANDLE_PD_RESUME,
```
`/** Handle power domain resume */`

gmarull commented on include/pm/device.h:
```diff
@@ -77,6 +80,12 @@ enum pm_device_action {
 	PM_DEVICE_ACTION_FORCE_SUSPEND,
+	/** Power domain resume */
+	PM_DEVICE_ACTION_HANDLE_PD_RESUME,
+	/** Power domain suspend */
+	PM_DEVICE_ACTION_HANDLE_PD_SUSPEND,
+	/** Power domain off */
+	PM_DEVICE_ACTION_HANDLE_PD_OFF,
```
ditto: `Handle ...`

gmarull commented on tests/subsys/pm/power_domain/src/main.c:
```diff
@@ -0,0 +1,151 @@
+	enum pm_device_state state;
+
+	domain = DEVICE_DT_GET(TEST_DOMAIN);
+	zassert_not_null(domain, "Failed to get device");
+
+	deva = DEVICE_DT_GET(TEST_DEVA);
+	zassert_not_null(deva, "Failed to get device");
+
+	devb = DEVICE_DT_GET(TEST_DEVB);
+	zassert_not_null(devb, "Failed to get device");
```
note: these can never be null, so assertion is redundant here

gmarull commented on tests/subsys/pm/power_domain/src/main.c:
```diff
@@ -0,0 +1,151 @@
+	devb = DEVICE_DT_GET(TEST_DEVB);
+	zassert_not_null(devb, "Failed to get device");
+
+	pm_device_runtime_enable(domain);
+	pm_device_runtime_enable(deva);
+	pm_device_runtime_enable(devb);
```
nit: returns error code, may be asserted (or ignored with `(void)` if failure is not possible)

ceolin replied:
In this pr revision `pm_device_runtime_enable` is still a void function. Will change it when rebasing ..

gmarull commented on doc/glossary.rst:
```diff
@@ -75,6 +75,11 @@ Glossary of Terms
       including its core services, device drivers, network stack, and so on.
 
+   power domain
+      A power domain is a collection of is a collection of devices for
```
nit: duplicate `is a collection`?

gmarull commented on doc/guides/pm/power_domain.rst:
```diff
@@ -0,0 +1,156 @@
+            /* suspend the domain */
+            ...
+            break;
+        case PM_DEVICE_ACTION_TURN_OFF:
+            /* turn off the domain */
+            ...
+            break;
```
note: is turn off mandatory? when is it used?

ceolin replied:
It is not mandatory, actually any of these actions are mandatory. A device can ignore if they want.

Regarding the turn off, is the same problem we have with `PM_DEVICE_STATE_OFF`. If we have a state to turn off a device and a domain is a device we need to have a related action for this.
We had this discussion about `PM_DEVICE_STATE_OFF` in the past, I was in favor of removing it but others wanted to keep it ... Having this state justify to have the domain action related to this.

gmarull commented on subsys/pm/device.c:
```diff
@@ -25,12 +25,25 @@ const char *pm_device_state_str(enum pm_device_state state)
+	return 0;
+}
+
 int pm_device_state_set(const struct device *dev,
 			enum pm_device_state state)
```
I think we should not update deprecated functions (`pm_device_state_set`).

ceolin replied:
reasons ?

ceolin replied:
fair enough

gmarull commented on subsys/pm/device.c:
```diff
@@ -84,6 +126,7 @@ int pm_device_action_run(const struct device *dev,
 	int ret;
 	enum pm_device_state state;
 	struct pm_device *pm = dev->pm;
+	bool domain;
```
note: this function is used by system PM. Consider this scenario: PD -> DEV during suspension.

- DEV is suspended
- PD is suspended. DEV is a dependent, so it will be notified with `PM_DEVICE_ACTION_HANDLE_PD_SUSPEND`. However, the device is suspended.

So my question: if I'm a device, under which conditions I'll receive a `PM_DEVICE_ACTION_HANDLE_PD_SUSPEND`? If I can get that notification when I'm both suspended and active, the actions I can/would take are likely different, so this needs some clarification IMO.

ceolin replied:
It is up to device decide what to do in this case, they have their own particular behavior, we cannot stipulate what they should do. What can be done is to make explicit that this workflow happens and that they should be aware.

gmarull commented:
@JordanYates review ping

ceolin commented:
@gmarull I've addressed most comments and replied another ones. Regarding error handling a pm state notification, I had replied in a previous version but basically my concerns are:

```
The problem is what do with another error.

If the domain is resuming the action happens before notify devices, in this case, if a device returns an error we would need to suspend the domain again ? And notify all devices that were already notified ?

If the domain is suspending, what to do, cancel the operation and notify devices that received the notification that the domain will continue active ?

Also, this will require all devices properly return -ENOTSUP or 0 when not handling these notifications. I don't know if initially we should be doing it, but I can change it though.
```

This inconsistency is something that can happen right now AFAIK since dependency between devices is not automatically handled, nothing prevents a device that supports be suspended when its dependents are still active.

gmarull commented on doc/guides/pm/power_domain.rst:
```diff
@@ -0,0 +1,161 @@
+                               enum pm_device_action *action)
+    {
+        switch (action) {
+        case PM_DEVICE_ACTION_SUSPEND:
+            /* suspend the domain */
+            ...
+            break;
+        case PM_DEVICE_ACTION_TURN_OFF:
```
I'm still failing to see when a domain will need to handle `PM_DEVICE_ACTION_SUSPEND` and `PM_DEVICE_ACTION_TURN_OFF`. What is the usecase for this added complexity (it has an associated `PM_DEVICE_ACTION_HANDLE_PD_OFF`)? I understand that domains can be either turned on (resumed in this language) or turned off (suspended).

JordanYates commented on include/pm/device.h:
```diff
@@ -77,6 +80,12 @@ enum pm_device_action {
 	/** Force suspend. */
 	PM_DEVICE_ACTION_FORCE_SUSPEND,
+	/** Handle power domain resume */
+	PM_DEVICE_ACTION_HANDLE_PD_RESUME,
```
I don't believe that it is necessary to create 3 new actions to support all the transitions that are possible.
Ignoring the `PM_DEVICE_STATE_SUSPENDING` state, there are 3 stable states for devices to exist in.

1. `PM_DEVICE_STATE_ACTIVE`
2. `PM_DEVICE_STATE_SUSPENDED`
3. `PM_DEVICE_STATE_OFF`

The actions `PM_DEVICE_ACTION_SUSPEND` and `PM_DEVICE_ACTION_RESUME` handle transitions between 1 and 2.
`PM_DEVICE_ACTION_TURN_OFF` handles 2 -> 3, so IMO we only need a `PM_DEVICE_ACTION_TURN_ON` to handle 3 -> 2.

ceolin replied:
That makes sense, the two proposed actions were not related to devices but just a way to notify about power domain transitions. I think just adding `TURN_ON` action is a better solution indeed.

JordanYates reviewed (commented):
I'm still strongly in favour of an example implementation for external devices and power domains. Like @gmarull, I also am not a fan of pushing logic into `pm_device_action_run`.

I will have a go at creating an alternative implementation to compare against.

Commit:
pm: device: add action to turn device on

Adds `PM_DEVICE_ACTION_TURN_ON` to transition from `PM_DEVICE_STATE_OFF`
to `PM_DEVICE_STATE_SUSPENDED`.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
pm: device: helper for running actions on children

Adds a helper function for running PM actions on child devices. A custom
implementation of `device_supported_foreach` is used as we don't
necessarily want the early exit behaviour of that function in the power
domain context.

The early exit behaviour can be obained through a return value in the
failure callback if that is desired.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
pm: device: helper to query power domain

Adds a helper for devices to check if they are on a power-domain or not.
Drivers can use this information to determine if they will be turned on
at some point in the future, or if they are already turned on.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
pm: device_runtime: init into PM_DEVICE_STATE_OFF

Add a function to tell runtime power management that the device is
starting in the off state instead of active or suspended.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
drivers: power_domain: gpio controlled domain

Initial implementation of a simple GPIO controlled power domain.
It exposes no API of its own, all functionality is contained inside
the runtime power management callbacks.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
tests: pm: power_domains: initial demo

Add an initial demonstration of a physical power domain running on the
`qemu_cortex_m3` platform.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
codeowners: Add power domains to the list

Add power domains directory and set it to myself.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: Properly handle -ENOTSUP

When a device returns -ENOTSUP for actions TURN_ON or TURN_OFF
the device state still has to be updated since the domain will
cut or restore the energy.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: Additional note about on/off actions

Add a note to clarify that these actions are only used by power
domains to notify children.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

gmarull commented on subsys/pm/device_runtime.c:
```diff
@@ -61,7 +61,7 @@ static int runtime_suspend(const struct device *dev, bool async)
 	} else {
 		/* suspend now */
-		ret = pm->action_cb(pm->dev, PM_DEVICE_ACTION_SUSPEND);
+		ret = pm_device_action_run(pm->dev, PM_DEVICE_ACTION_SUSPEND);
```
this patch should not be required now, device_runtime can easily stay away from `pm_device_action_run`.

ceolin commented on subsys/pm/device.c:
```diff
@@ -119,6 +119,13 @@ int pm_device_action_run(const struct device *dev,
 		break;
+	case PM_DEVICE_ACTION_TURN_ON:
+		if (pm->state != PM_DEVICE_STATE_OFF) {
+			return -EALREADY;
```
good catch.

nashif commented on tests/subsys/pm/device_power_domains/prj.conf:
```diff
@@ -0,0 +1,10 @@
+CONFIG_ZTEST=y
```
when did we ever put copyrights in conf files?

ceolin replied:
The majority of  these files don't have copyright.  Just a few have, since it does not hurt I'll add though.

ceolin commented on subsys/pm/device.c:
```diff
@@ -140,6 +140,36 @@ int pm_device_action_run(const struct device *dev,
+		rc = pm_device_action_run(cdev, action);
+		if (failure_cb && rc < 0) {
+			/* Stop the iteration if the callback requests it */
+			if (!failure_cb(cdev, rc)) {
```
nope, this function returns a bool.

ceolin commented on tests/subsys/pm/power_domain/src/main.c:
```diff
@@ -0,0 +1,212 @@
+	case PM_DEVICE_ACTION_SUSPEND:
+		pm_device_children_action_run(dev, PM_DEVICE_ACTION_TURN_OFF, NULL);
+		break;
+	case PM_DEVICE_ACTION_TURN_ON:
+		__fallthrough;
+	case PM_DEVICE_ACTION_TURN_OFF:
+		break;
```
It does not really matter, this is the action for the domain and the test don't have nested power domain so these actions are not called. The test though, has two devices, one handles `TURN_ON/OFF` the other returns `-ENOTSUP` for these actions and the test checks if the state is consistent.

gmarull commented on doc/guides/pm/device.rst:
```diff
@@ -134,3 +134,12 @@ later calling :c:func:`pm_device_wakeup_enable`.
+
+Power domain on Zephyr is represented as a regular device. The power management
+subsystem ensures that a domain is resumed before and suspended after devices
+using it. When ::kconfig:`CONFIG_PM_DEVICE_RUNTIME` is enabled, every time a
```
nit: `::kconfig` -> `:kconfig:`

gmarull commented on subsys/pm/device.c:
```diff
@@ -141,8 +141,8 @@ int pm_device_action_run(const struct device *dev,
 void pm_device_children_action_run(const struct device *dev,
-		enum pm_device_action action,
-		pm_device_action_failed_cb_t failure_cb)
+				   enum pm_device_action action,
+				   pm_device_action_failed_cb_t failure_cb)
```
ping

gmarull commented on drivers/power_domain/power_domain_gpio.c:
```diff
@@ -0,0 +1,109 @@
+	case PM_DEVICE_ACTION_RESUME:
+		/* Switch power on */
+		gpio_pin_set_dt(&cfg->enable, 1);
+		LOG_INF("%s is now ON", dev->name);
+		/* Notify supported devices they are now powered */
+		pm_device_children_action_run(dev, PM_DEVICE_ACTION_TURN_ON, NULL);
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		/* Notify supported devices power is going down */
+		pm_device_children_action_run(dev, PM_DEVICE_ACTION_TURN_OFF, NULL);
+		/* Switch power off */
+		gpio_pin_set_dt(&cfg->enable, 0);
+		LOG_INF("%s is now OFF and powered", dev->name);
+		break;
+	case PM_DEVICE_ACTION_TURN_ON:
+		/* Actively control the enable pin now that the device is powered */
+		gpio_pin_configure_dt(&cfg->enable, GPIO_OUTPUT_INACTIVE);
+		LOG_INF("%s is OFF and powered", dev->name);
+		break;
+	case PM_DEVICE_ACTION_TURN_OFF:
+		/* Let the enable pin float while device is not powered */
+		gpio_pin_configure_dt(&cfg->enable, GPIO_DISCONNECTED);
+		LOG_INF("%s is OFF and not powered", dev->name);
```
nit: should be `LOG_DBG`

gmarull reviewed (approved):
Major request: move TI driver to `drivers/power_domain` (please do in this PR since change should be pretty self-contained)
